Feature/modify Audio Streams and Vibration.

francescobau opened the pull request:
Hello there.
This is my utility app to manage Audio Streams and Vibration.
Though I'm a bit concerned about 2 things:
- **Bad Android-Testing**. I don't know how am I supposed to test all methods, especially the ones involving vibration.
- **Bad way to manage Context**. This is an issue that the PO also confirmed. How am I supposed to manage the Context in a better way, even if I use only static methods? The easiest way is to make the methods as non-static, and use an object to use my utility class, passing the Context in the constructor, but, theoretically, someone would be able to have 2 objects to manage Volume and Vibration... also, I think the use of static methods is more user-friendly. Do you have any suggestions? Any help would be very appreciated.
Thank you in advance.

Real changes:
- added AudioUtilityManager , the utility class
- added AudioUtilityManagerTest, the android-test, with suite test as well
- updated AndroidManifest.xml , to add permissions (Audio, Vibration and Notifications)

francescobau committed:
Update from remote - to fix syncronization problems

Here are the git commands to do when encountering troubles with Android Studio:
1) git submodule update --remote --recursive
2) git submodule update --remote --merge

francescobau committed:
Created suite test for audioUtility, added Audio, Notifications and Vibration permissions in Manifest, imported classes from old branch, updated resources from old branch.

francescobau committed:
Cleared useless imports from AudioUtilityManagerTest

These imports used to be important in the past, but now they are not necessary, because I've created a suite test with CTRL+SHIFT+T .

francescobau committed:
Library update?

I created a branch from previous commit in my branch, and it asked me in the master to commit. In doubt, I discarded. But it keeps to appear, so I'm gonna do this in my branch. If necessary, people can come and get these codes.

francescobau committed:
Cleared MainActivity, restored old resource folder.

francescobau committed:
Update library

francescobau committed:
added import android.os.Bundle , which is necessary.

francescobau committed:
Easy drag of import, to have -1 files changed

CremaLuca commented on .idea/misc.xml:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_7" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
We decided to keep java 8 in smslibrary, I think we should be consistent with that choice and use 8

francescobau replied:
Thank you for pointing this out.
Will fix that soon.

CremaLuca commented on app/src/androidTest/java/com/eis/audioUtility/AudioUtilityManagerTest.java:
```diff
@@ -0,0 +1,202 @@
+            AudioUtilityManager.setVolume(context, AudioUtilityManager.AUMStream.ALARM, -1);
+            Assert.fail("Should have thrown IllegalArgumentException.");
+        } catch (IllegalArgumentException e) {
+            //Success
```
IllegalArgumentException is a runtime exception, you can remove all of this an keep the `AudioUtilityManager.setVolume(...)` line and add on top of the method `@Test(excepted = IllegalArgumentException.class)`, it will do this exact same job

francescobau replied:
Ok. Thank you for the advice.

CremaLuca commented on app/src/androidTest/java/com/eis/audioUtility/AudioUtilityManagerTest.java:
```diff
@@ -0,0 +1,202 @@
+        try {
+            AudioUtilityManager.getVolume(context, AudioUtilityManager.AUMStream.ALARM);
+            //Success
+        } catch (Exception e) {
```
Test fails automatically if an exception is thrown

francescobau replied:
ok.. So what am I supposed to do? Should I delete this test?

CremaLuca replied:
Yes, you can delete the try-catch and just leave the AudioUtility line

francescobau replied:
Should I delete all these methods with catch(Exception e) declaration?
[Those ones with catch(IllegalArgumentException e) as well]

e.g.

@ Test
    public void setMusicVolumeToMax_maxVolume_isOK() {
        try {
            AudioUtilityManager.setVolumeToMax(context, AudioUtilityManager.AUMStream.MUSIC);
            //Success
        } catch (IllegalArgumentException e) {
            Assert.fail(e.getStackTrace() + "\nMAX_PERCENTAGE is NOT OK. Please check this value.");
        } **catch (Exception x) {
            Assert.fail(x.getStackTrace() + "\nIt's NOT supposed to fail like this!");**
        }
    }

CremaLuca commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,199 @@
+    /**
+     * @param context The current Context.
+     * @param stream  The chosen stream (should be ALARM, RING or MUSIC).
+     * @return The maximum stream volume (real value).
```
Specify that this value is not always a percentage

francescobau replied:
In fact, it's not a percentage, it's the real value.
So should I write a better description of this method?
This one is important, because it's needed to calculate the percentage...

CremaLuca replied:
You should say in the @return that the maximum stream volume is not a percentage

francescobau replied:
ok, thank you for the advice.

CremaLuca commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,199 @@
+
+    /**
+     * @param context The current Context.
+     * @param stream  The chosen stream (should be ALARM, RING or MUSIC).
```
Saying what it should be is redundant, those are the only values you can choose

francescobau replied:
Ok, I'll fix that. Thanks.

CremaLuca commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,199 @@
+        getAudioManager(context).setStreamVolume(
+                getStream(stream),
+                newVolume,
+                AudioManager.FLAG_SHOW_UI
```
What if I didn't want to have the user see that I'm modifying the audio volume? You could add a parameter like `notifyUser` that enables or disables this show on UI

francescobau replied:
That's a good idea!
Is that supposed to be a boolean, then?

CremaLuca replied:
Idk, see AudioManager documentation, it's probably a boolean

francescobau replied:
Not really, the flag is not a boolean... it's an int value.
Sorry for my bad English, what I meant to ask you is: should I add a new boolean parameter in the method? So, if true, the flag would be shown, and it would be hidden otherwise.

But we get into another issue by doing this: there are gonna be 4 parameters in this method...
I was thinking about an overload of the method, but the only one I can think about is to set the volume with the default Stream... And, in our app, it should be Alarm, right?

CremaLuca commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,199 @@
+     */
+    public static void setVolume(Context context, AUMStream stream, int percentage) throws IllegalArgumentException {
+        if (percentage < MIN_PERCENTAGE)
+            throw new IllegalArgumentException(" Your value is too low. Please insert a value between 0 and 100.");
```
Is there a reason why the exception message starts with a space?

francescobau replied:
Oh shoot, I got used by e.getStackTrace(), which printed the default IllegalArgumentException (sorry for my bad explanation), and, if there's no space, they get stuck together. I should have used \n instead.
But I guess it's useless in this situation because in the error Logcat we can see only the string, right? Or it prints the default one as well?
BTW. thank you for pointing this out.

CremaLuca commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,199 @@
+            case MUSIC:
+                return AudioManager.STREAM_MUSIC;
+            default:
+                throw new IllegalArgumentException("\nIllegal Stream. It should be ALARM, RING or MUSIC .");
```
Also, is there a reason why this exception starts with a new line command?

francescobau replied:
See the comment above.
I guess I'm gonna remove it.

CremaLuca commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,199 @@
+     * @param context The current Context.
+     */
+    public static void stopVibrate(Context context) {
+        Vibrator vibrator = (Vibrator) context.getSystemService(Context.VIBRATOR_SERVICE);
```
Should have used the `getVibrator()` method

francescobau replied:
Oh, you are totally right!
Thank you for pointing this out.

CremaLuca commented on app/src/main/java/com/eis/phoneringer/MainActivity.java:
```diff
@@ -4,11 +4,14 @@
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
+
```
All you did to this file was... Adding spaces? You should revert these changes in order to have a cleaner pull request

francescobau replied:
Not really, I cleared my MainActivity and reformatted it, but it looks like the MainActivity in the main repository is NOT formatted...

CremaLuca commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,199 @@
+
+import static android.content.Context.AUDIO_SERVICE;
+
+
```
What about a class description?

CremaLuca replied:
Oh wait it's up there, you should move it attached to the class name, on this line

francescobau replied:
Ok will do it. Thank you.

francescobau committed:
Updated JDK to 1.8, comments and specification fixing (see description below), edited method stopVibrate(Context), android-test optimization,

- Both classes' descriptions are dragged near class declaration
- some redundant comments, such as "should be " have been optimized
- edited method stopVibrate(Context)
- optimized android-testing, especially for exceptions
KNOWN ISSUES:
It's still not done.

francescobau committed:
Finished android-test optimization for Stream Volumes + resolved some hardwiring issues

- added expected after @ Test declaration, so optimized abuse of try-catch
- resolved hardwiring issue with the percentage, since that's redundant

francescobau committed:
General optimization of AudioUtilityManager class: optimized Exceptions usage.

francescobau committed:
Edited method setVolume , adding boolean as a parameter; added Overload of method setVolume(Context,AUMStream,int,boolean). Added @ NonNull declarations, and fixed specifications.

KNOWN ISSUES:
- still unable to test Vibration methods

francescobau committed:
Changed project labguageLevel in JDK_1_8

francescobau committed:
Added PermissionDialog, attempt to force JDK version to 1.8

DISCLAIMER: still under development.

francescobau committed:
Removed PermissionDialog from SA

Cogno-Marco commented on app/src/androidTest/java/com/eis/audioUtility/AudioUtilityManagerTest.java:
```diff
@@ -0,0 +1,167 @@
+    @Test(expected = IllegalArgumentException.class)
+    public void setAlarmVolume_percentage_isTooLow() {
+        AudioUtilityManager.setVolume(context, AudioUtilityManager.AUMStream.ALARM, INVALID_PERCENTAGE_LOW);
+        Assert.fail("Should have thrown IllegalArgumentException.");
```
Little suggestion, since there's the `expected = IllegalArgumentException.class`, the `fail()` is redundant. This test will automatically fail if no exception is thrown and also if the wrong exception gets thrown

francescobau replied:
You are totally right! Thank you, I'm gonna fix it right away!

Cogno-Marco commented on app/src/androidTest/java/com/eis/audioUtility/AudioUtilityManagerTest.java:
```diff
@@ -0,0 +1,167 @@
+            Assert.fail(e.getStackTrace() + "\nMAX_PERCENTAGE is NOT OK. Please check this value.");
+        } catch (Exception x) {
+            Assert.fail(x.getStackTrace() + "\nIt's NOT supposed to fail like this!");
+        }
```
Also here, any exception is automatically a test failed, you don't need to block it and make the test fail yourself

francescobau replied:
Thank you again.

Cogno-Marco commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * @param context The current Context
```
I get that the name is obvious etc. but I would have added some more documentation anyway. Something like
`Returns the current AudioManager instance`

francescobau replied:
Thank you, I've fixed it.

Cogno-Marco commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * @param context The current Context.
```
Same as line 38

Cogno-Marco commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * @param context The current Context.
```
Same as line 38 too

Cogno-Marco commented on app/src/main/java/com/eis/audioUtility/AudioUtilityManager.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * @param context The current Context.
```
Same as line 38

Cogno-Marco reviewed (changes requested):
Requesting a couple of test code improvements and documentation/specifications improvements

francescobau committed:
added specifications, fixed some Match Case issues, cleaned redoundant Assert.fail in test class.

KNOWN ISSUES:
- Still don't know how to test Vibration methods
- Method "singleVibrate" uses a default time, which is not customizable.

francescobau commented:
Ok I think I've improved it...
@Cogno-IDU does everything look better, now?

Cogno-Marco reviewed (approved):
Yup, everything's fine now 👍
